@here/harp-mapview/lib/PickListener.ts

```diff
@@ -0,0 +1,127 @@
+ */
+
+import { IntersectParams } from "./IntersectParams";
+import { PickResult } from "./PickHandler";
+
+// Default sorting by distance first and then by reversed render order.
+function defaultSort(lhs: PickResult, rhs: PickResult) {
+    const distanceDiff = lhs.distance - rhs.distance;
+    const haveRenderOrder = lhs.renderOrder !== undefined && rhs.renderOrder !== undefined;
+    if (distanceDiff !== 0 || !haveRenderOrder) {
+        return distanceDiff;
+    }
+
+    return rhs.renderOrder! - lhs.renderOrder!;
+}
+
+/**
+ * Collects results from a picking (intersection) test.
+ *
+ * @internal
+ */
+export class PickListener {
+    private m_results: PickResult[] = [];
+    private m_sorted: boolean = true;
+
+    /**
+     * Constructs a new `PickListener`.
+     *
+     * @param m_parameters - Optional parameters to customize picking behaviour.
+     */
+    constructor(private readonly m_parameters?: IntersectParams) {}
+
+    /**
+     * Adds a pick result.
+     *
+     * @param result - The result to be added.
+     */
+    addResult(result: PickResult): void {
+        // Add the result only if it's a different feature from the ones already collected.
+        const foundFeatureIdx = this.m_results.findIndex(otherResult => {
+            const sameType = otherResult.type === result.type;
+            const dataSource = result.intersection?.object.userData?.dataSource;
+            const sameDataSource =
+                dataSource && otherResult.intersection?.object.userData?.dataSource === dataSource;
+            const sameId =
+                result.featureId !== undefined && otherResult.featureId === result.featureId;
+            const noId = result.featureId === undefined && otherResult.featureId === undefined;
+            const sameUserData = result.userData && otherResult.userData === result.userData;
+            return sameType && sameDataSource && (sameId || (noId && sameUserData));
+        });
+
+        if (foundFeatureIdx < 0) {
+            this.m_sorted = false;
+            this.m_results.push(result);
+            return;
+        }
+
+        // Replace the result for the same feature if it's sorted after the new result.
+        const oldResult = this.m_results[foundFeatureIdx];
+        if (defaultSort(result, oldResult) < 0) {
+            this.m_results[foundFeatureIdx] = result;
+            this.m_sorted = false;
+        }
+    }
+
+    /**
+     * Indicates whether the listener is satisfied with the results already provided.
+     * @returns `True` if the listener doesn't expect more results, `False` otherwise.
+     */
+    get done(): boolean {
+        return this.maxResults ? this.m_results.length >= this.maxResults : false;
+    }
+
+    /**
+     * Returns the collected results, ordered by distance first, then by reversed render order
+     * (topmost/highest render order first).
+     * @returns The pick results.
+     */
+    get results(): PickResult[] {
+        this.finish();
```

Comment:
I would rather do an explicit call to finish in the pick handler instead of having side effects in the getter

Reply:
yes, makes sense, will fix it in a separate PR.